src/Solnet.Rpc/SolanaRpcBatchComposer.cs

```diff
@@ -0,0 +1,384 @@
+        /// </summary>
+        /// <param name="input"></param>
+        /// <param name="nativeType"></param>
+        /// <returns></returns>
+        public object MapJsonTypeToNativeType(object input, Type nativeType)
+        {
+            if (input is JsonElement)
+            {
+                // serializes + deserializes the JSON into runtime type - suboptimal but expedient
+                var elem = (JsonElement)input;
+                var bufferWriter = new ArrayBufferWriter<byte>();
+                using (var writer = new Utf8JsonWriter(bufferWriter))
+                    elem.WriteTo(writer);
+                return JsonSerializer.Deserialize(bufferWriter.WrittenSpan, nativeType, _jsonOptions);
+            }
+            else
+            {
+                return input;
+            }
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Clears the internal list of requests
+        /// </summary>
+        public void Clear()
+        {
+            _reqs.Clear();
+        }
+
+        /// <summary>
+        /// Executes any batch using the auto execution mode (if set) or throws an execption.
+        /// </summary>
+        public void Flush()
+        {
+            switch (_autoMode)
+            {
+                case BatchAutoExecuteMode.ExecuteWithFatalFailure:
+                    ExecuteWithFatalFailure(_autoRpcClient);
+                    break;
+
+                case BatchAutoExecuteMode.ExecuteWithCallbackFailures:
+                    Execute(_autoRpcClient);
+                    break;
+
+                default:
+                    throw new ApplicationException("BatchComposer AutoExecute mode not set");
+            }
+        }
+
+        internal void AddRequest<T>(string method, IList<object> parameters, Action<T, Exception> callback)
+        {
+            var wrapped = WrapCallback<T>(callback);
+            var handler = RpcBatchReqRespItem.Create<T>(_idGenerator.GetNextId(), method, parameters, wrapped);
+            Add(handler);
+        }
+
+        internal Task<T> AddRequest<T>(string method, IList<object> parameters)
+        {
+            var taskSource = new TaskCompletionSource<T>();
+            var callback = WrapTaskSource<T>(taskSource);
+            var handler = RpcBatchReqRespItem.Create<T>(_idGenerator.GetNextId(), method, parameters, callback);
+            Add(handler);
+            return taskSource.Task;
+        }
+
+        internal void Add(RpcBatchReqRespItem task) 
+        {
+            // add to batch
+            _reqs.Add(task);
+
+            // does this trigger an auto execute?
+            if (_autoMode != BatchAutoExecuteMode.Manual && _reqs.Count >= _autoBatchSize)
+                Flush();
+        }
+
+        private static Action<JsonRpcBatchResponseItem, Exception> WrapCallback<T>(Action<T, Exception> callback)
+        {
+            if (callback == null) return null;
```

Comment:
Question, wouldn't it be more clean to have a callback for data, and a optional callback for exception handling?
Would also make consumer code cleaner and easier to reason about, as you don't need to distinguish between calls for exception handling and calls for data.

Reply:
I can see that some developers may want separate success/failure callbacks and depending on the complexity of the calling code. 

This style with combined `(data, ex)` args has been used by several other developers including myself and nobody came back with this suggestion when asking for feedback. I suggest we ship what we have, and if the wider communitity would like to experiement with this separate callback style we can introduce a new `SolanaRpcBatchWithDualCallbacks` as an RFC like we tried and ultimately killed off with the `SolanaRpcBatchWithAsyncs` experiements.